File: 0731-my-calendar-ii/0731-my-calendar-ii.py
class MyCalendarTwo:
    def __init__(self):
        self.cal = []  
        self.overlap = []  

    def overlapped(self, start1, end1, start2, end2):
        return max(start1, start2) < min(end1, end2)

    def overlapped_interval(self, start1, end1, start2, end2):
        return (max(start1, start2), min(end1, end2))

    def book(self, start, end):
        for book in self.overlap:
            if self.overlapped(book[0], book[1], start, end):
                return False  


        for book in self.cal:
            if self.overlapped(book[0], book[1], start, end):
                self.overlap.append(self.overlapped_interval(book[0], book[1], start, end))


        self.cal.append((start, end))
        return True 

# Example usage:
# obj = MyCalendarTwo()
# param_1 = obj.book(start, end)

File: 0731-my-calendar-ii/0731-my-calendar-ii.java
import java.util.ArrayList;
import java.util.List;

class MyCalendarTwo {
    private List<int[]> cal;
    private List<int[]> overlap;

    public MyCalendarTwo() {
        cal = new ArrayList<>();
        overlap = new ArrayList<>();
    }

    private boolean overlapped(int start1, int end1, int start2, int end2) {
        return Math.max(start1, start2) < Math.min(end1, end2);
    }

    private int[] overlappedInterval(int start1, int end1, int start2, int end2) {
        return new int[]{Math.max(start1, start2), Math.min(end1, end2)};
    }

    public boolean book(int start, int end) {
        for (int[] book : overlap) {
            if (overlapped(book[0], book[1], start, end)) {
                return false;  
            }
        }
        for (int[] book : cal) {
            if (overlapped(book[0], book[1], start, end)) {
                overlap.add(overlappedInterval(book[0], book[1], start, end));
            }
        }

        cal.add(new int[]{start, end});
        return true; 
    }
}

/**
 * Your MyCalendarTwo object will be instantiated and called as such:
 * MyCalendarTwo obj = new MyCalendarTwo();
 * boolean param_1 = obj.book(start, end);
 */
